File: ios/PlaytomicFoundation/promise/Promise.swift
//
//  Promise.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 27/01/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation
import PromiseKit

public class Promise<T> {
    var promise: PromiseKit.Promise<T>

    public required init(
        after: TimeInterval = 0,
        executeInBackground: Bool = false,
        resolvers: @escaping (_ fulfill: @escaping (T) -> Void, _ reject: @escaping (Error) -> Void) throws -> Void
    ) {
        self.promise = PromiseKit.Promise { resolver in
            Executor.execute(after: after, inBackground: executeInBackground) {
                do {
                    try resolvers(resolver.fulfill, resolver.reject)
                } catch {
                    resolver.reject(error)
                }
            }
        }
    }

    public required init(value: T) {
        self.promise = PromiseKit.Promise.value(value)
    }

    public required init(error: Error) {
        self.promise = PromiseKit.Promise(error: error)
    }

    internal init(promise: PromiseKit.Promise<T>) {
        self.promise = promise
    }

    public var value: T? {
        promise.value
    }

    public var error: Error? {
        promise.error
    }

    public var isResolved: Bool {
        value != nil || error != nil
    }

    @discardableResult
    public func then(inBackground: Bool = false, _ body: @escaping (T) throws -> Void) -> Promise<T> {
        then(inBackground: inBackground, map: { value in
            try body(value)
            return value
        })
    }

    @discardableResult
    public func then<U>(inBackground: Bool = false, map body: @escaping (T) throws -> U) -> Promise<U> {
        let queue = inBackground ? DispatchQueue.global(qos: .background) : DispatchQueue.main
        return Promise<U>(promise: promise.map(on: queue, body))
    }

    @discardableResult
    public func then<U>(inBackground: Bool = false, promise body: @escaping (T) throws -> Promise<U>) -> Promise<U> {
        let (pendingPromise, resolver) = PromiseKit.Promise<U>.pending()
        promise
            .done { (result: T) in
                Executor.execute(inBackground: inBackground) {
                    do {
                        let deferredPromise = try body(result)
                        deferredPromise
                            .then(map: resolver.fulfill)
                            .catchError(resolver.reject)
                    } catch {
                        resolver.reject(error)
                    }
                }
            }.catch(resolver.reject)

        return Promise<U>(promise: pendingPromise)
    }

    @discardableResult
    public func catchError(_ body: @escaping (Error) -> Void) -> Promise<T> {
        promise.catch { error in
            body(error)
        }
        return Promise<T>(promise: promise)
    }

    @discardableResult
    public func fulfillOnError(value: T) -> Promise<T> {
        Promise<T> { fulfill, _ in
            self.then(fulfill)
                .catchError { _ in fulfill(value) }
        }
    }

    @discardableResult
    public func fulfillOnError(promise errorHandler: @escaping (Error) -> Promise<T>) -> Promise<T> {
        Promise<T> { fulfill, reject in
            self.then(fulfill)
                .catchError { error in
                    errorHandler(error).then(fulfill).catchError(reject)
                }
        }
    }

    @discardableResult
    public func fulfillOnError(map errorHandler: @escaping (Error) -> T?) -> Promise<T> {
        Promise<T> { fulfill, reject in
            self.then(fulfill)
                .catchError { error in
                    if let value = errorHandler(error) {
                        fulfill(value)
                    } else {
                        reject(error)
                    }
                }
        }
    }

    @discardableResult
    public func always(inBackground: Bool = false, _ body: @escaping () -> Void) -> Promise<T> {
        let queue = inBackground ? DispatchQueue.global(qos: .background) : DispatchQueue.main
        return Promise<T>(promise: promise.ensure(on: queue, body))
    }

    @discardableResult
    public func rejectOnTimeout(timeout: Double) -> Promise<T> {
        let (pendingPromise, resolver) = PromiseKit.Promise<T>.pending()
        let task = Executor.execute(after: timeout, inBackground: false) {
            if pendingPromise.isPending {
                resolver.reject(FoundationException.timeout)
            }
        }
        promise
            .done { (result: T) in
                if pendingPromise.isPending {
                    task?.cancel()
                    resolver.fulfill(result)
                }
            }
            .catch {
                if pendingPromise.isPending {
                    task?.cancel()
                    resolver.reject($0)
                }
            }

        return Promise(promise: pendingPromise)
    }

    @discardableResult
    public func ignoreResult() -> Promise<Void> {
        then(map: { _ in })
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/promise/Promise.kt
package com.playtomic.foundation.promise

import com.playtomic.foundation.executor.Executor
import com.playtomic.foundation.extension.tryOrNull
import com.playtomic.foundation.model.FoundationException
import org.jdeferred.android.AndroidDeferredManager
import org.jdeferred.impl.DeferredObject

/**
 * Created by mgonzalez on 20/12/16.
 */

open class Promise<D> {
    companion object {
        val mainManager = AndroidDeferredManager()
    }

    internal val promise: org.jdeferred.Promise<D, Exception, Double>

    var value: D? = null

    var error: Exception? = null

    val isResolved: Boolean
        get() = value != null || error != null

    constructor(value: D) {
        val deferred = DeferredObject<D, Exception, Double>()
        promise = mainManager.`when`(deferred.promise()).then { this.value = it }.fail { this.error = it }
        deferred.resolve(value)
    }

    constructor(error: Exception) {
        val deferred = DeferredObject<D, Exception, Double>()
        promise = mainManager.`when`(deferred.promise()).then { this.value = it }.fail { this.error = it }
        deferred.reject(error)
    }

    constructor(after: Double = 0.0, executeInBackground: Boolean = false, resolver: (fulfill: (D) -> Unit, reject: (Exception) -> Unit) -> Unit) {
        val deferred = DeferredObject<D, Exception, Double>()
        promise = mainManager.`when`(deferred).then { this.value = it }.fail { this.error = it }
        Executor.execute(after = after, inBackground = executeInBackground) {
            try {
                resolver({
                    if (!deferred.isResolved) {
                        deferred.resolve(it)
                    }
                }, {
                    if (!deferred.isResolved) {
                        deferred.reject(it)
                    }
                })
            } catch (e: Exception) {
                if (!deferred.isResolved) {
                    deferred.reject(e)
                }
            }
        }
    }

    protected constructor(promise: org.jdeferred.Promise<D, Exception, Double>) {
        this.promise = mainManager.`when`(promise).then { this.value = it }.fail { this.error = it }
    }

    fun catchError(callback: (Exception) -> Unit): Promise<D> =
        Promise<D>(promise.fail(callback))

    fun fulfillOnError(promise: (Exception) -> Promise<D>): Promise<D> =
        Promise { fulfill, reject ->
            this.then(fulfill)
                .catchError { error ->
                    promise(error)
                        .then(fulfill)
                        .catchError(reject)
                }
        }

    fun fulfillOnError(value: D): Promise<D> =
        Promise { fulfill, _ ->
            this
                .then(fulfill)
                .catchError { fulfill(value) }
        }

    @JvmName("fulfillOnErrorWithNullable")
    fun fulfillOnError(map: (Exception) -> D?): Promise<D> =
        Promise { fulfill, reject ->
            this.then(fulfill).catchError { error ->
                val value = tryOrNull { map(error) }
                if (value != null) {
                    fulfill(value)
                } else {
                    reject(error)
                }
            }
        }

    fun then(execute: (D) -> Unit): Promise<D> =
        then(inBackground = false, execute = execute)

    fun then(inBackground: Boolean, execute: (D) -> Unit): Promise<D> {
        val deferred = DeferredObject<D, Exception, Double>()
        this.promise.then { result: D ->
            Executor.execute(inBackground = inBackground) {
                try {
                    execute(result)
                    deferred.resolve(result)
                } catch (e: Exception) {
                    deferred.reject(e)
                }
            }
        }.fail { deferred.reject(it) }
        return Promise(deferred)
    }

    @JvmName("thenMap")
    fun <U> then(map: (D) -> U): Promise<U> =
        then(inBackground = false, map = map)

    @JvmName("thenMapInBackground")
    fun <U> then(inBackground: Boolean, map: (D) -> U): Promise<U> {
        val deferred = DeferredObject<U, Exception, Double>()
        this.promise.then { result: D ->
            Executor.execute(inBackground = inBackground) {
                try {
                    val newObject = map(result)
                    deferred.resolve(newObject)
                } catch (e: Exception) {
                    deferred.reject(e)
                }
            }
        }.fail { deferred.reject(it) }
        return Promise(deferred)
    }

    @JvmName("thenPromise")
    fun <D_OUT> then(promise: (D) -> Promise<D_OUT>): Promise<D_OUT> =
        then(inBackground = false, promise = promise)

    @JvmName("thenPromiseInBackground")
    fun <D_OUT> then(inBackground: Boolean, promise: (D) -> Promise<D_OUT>): Promise<D_OUT> {
        val deferred = DeferredObject<D_OUT, Exception, Double>()
        this.promise
            .then { result: D ->
                Executor.execute(inBackground = inBackground) {
                    try {
                        val deferredPromise = promise(result)
                        deferredPromise
                            .then(deferred::resolve)
                            .catchError(deferred::reject)
                    } catch (e: Exception) {
                        deferred.reject(e)
                    }
                }
            }
            .fail { deferred.reject(it) }

        return Promise(deferred)
    }

    fun always(callback: () -> Unit): Promise<D> =
        always(inBackground = false, callback = callback)

    fun always(inBackground: Boolean, callback: () -> Unit): Promise<D> {
        // agarcia: We use then/fail instead of always because JDeferred executes always callback as last step and not in setup order as expected
        val deferred = DeferredObject<D, Exception, Double>()
        this.promise.then { result: D ->
            Executor.execute(inBackground = inBackground) {
                callback()
                deferred.resolve(result)
            }
        }.fail {
            Executor.execute(inBackground = inBackground) {
                callback()
                deferred.reject(it)
            }
        }
        return Promise(deferred)
    }

    fun rejectOnTimeout(timeout: Double): Promise<D> {
        val deferred = DeferredObject<D, Exception, Double>()
        val task = Executor.execute(after = timeout, inBackground = false) {
            if (deferred.isPending) {
                deferred.reject(FoundationException.timeout)
            }
        }
        promise
            .then { result: D ->
                if (deferred.isPending) {
                    task?.cancel()
                    deferred.resolve(result)
                }
            }
            .fail {
                if (deferred.isPending) {
                    task?.cancel()
                    deferred.reject(it)
                }
            }
        return Promise(deferred)
    }

    fun ignoreResult(): Promise<Unit> = then(map = {})
}
